ignite/engine/__init__.py

```diff
@@ -69,48 +287,71 @@ def create_supervised_trainer(
 
     Note:
         `engine.state.output` for this engine is defined by `output_transform` parameter and is the loss
         of the processed batch by default.
 
     .. warning::
-
         The internal use of `device` has changed.
         `device` will now *only* be used to move the input data to the correct device.
         The `model` should be moved by the user before creating an optimizer.
         For more information see:
 
         - `PyTorch Documentation <https://pytorch.org/docs/stable/optim.html#constructing-it>`_
-
         - `PyTorch's Explanation <https://github.com/pytorch/pytorch/issues/7844#issuecomment-503713840>`_
 
+    .. warning::
+        If ``amp_mode='apex'`` , the model(s) and optimizer(s) must be initialized beforehand
+        since ``amp.initialize`` should be called after you have finished constructing your model(s)
+        and optimizer(s), but before you send your model through any DistributedDataParallel wrapper.
+
+        See more: https://nvidia.github.io/apex/amp.html#module-apex.amp
+
     Returns:
         Engine: a trainer engine with supervised update function.
+
+    .. versionchanged:: 0.5.0
+
+        - Added ``amp_mode`` argument for automatic mixed precision.
+        - Added ``scaler`` argument for gradient scaling.
     """
 
     device_type = device.type if isinstance(device, torch.device) else device
     on_tpu = "xla" in device_type if device_type is not None else False
-
-    if on_tpu and not idist.has_xla_support:
-        raise RuntimeError("In order to run on TPU, please install PyTorch XLA")
-
-    def _update(engine: Engine, batch: Sequence[torch.Tensor]) -> Union[Any, Tuple[torch.Tensor]]:
-        model.train()
-        optimizer.zero_grad()
-        x, y = prepare_batch(batch, device=device, non_blocking=non_blocking)
-        y_pred = model(x)
-        loss = loss_fn(y_pred, y)
-        loss.backward()
-
-        if on_tpu:
-            xm.optimizer_step(optimizer, barrier=True)
-        else:
-            optimizer.step()
-
-        return output_transform(x, y, y_pred, loss)
+    mode = _check_arg(on_tpu, amp_mode, scaler)
+
+    if mode == "amp":
+        _update, scaler_ = supervised_training_step_amp(
+            model, optimizer, loss_fn, device, non_blocking, prepare_batch, output_transform, scaler
+        )
+    elif mode == "apex":
+        _update = supervised_training_step_apex(
+            model, optimizer, loss_fn, device, non_blocking, prepare_batch, output_transform
+        )
+    elif mode == "tpu":
+        _update = supervised_training_step_tpu(
+            model, optimizer, loss_fn, device, non_blocking, prepare_batch, output_transform
+        )
+    else:
+        _update = supervised_training_step(
+            model, optimizer, loss_fn, device, non_blocking, prepare_batch, output_transform
+        )
 
     trainer = Engine(_update) if not deterministic else DeterministicEngine(_update)
+    if scaler and isinstance(scaler, bool) and mode == "amp":
```

Comment:
```suggestion
    if _scaler:
```

Reply:
I am afraid that it would result in variable undefined in not scaler cases.

Reply:
of course, my mistake !

Reply:
Maybe we should include rules in `_check_arg` 
```
mode, _scaler = _check_arg(on_tpu, amp_mode, scaler) # build scaler here or None

if mode == "amp":
        _update = supervised_training_step_amp(
            model, optimizer, loss_fn, device, non_blocking, prepare_batch, output_transform, _scaler
        )
elif ...

trainer = Engine(_update) if not deterministic else DeterministicEngine(_update)

if _scaler:
    trainer.state.scaler = _scaler
```

What do you think ? I insist because line 358 is related to a rule used inside `supervised_training_step_amp`. Therefore, change this rule means modify 2 places in code...

Reply:
Yes, I agree, it's better API
Thank you!